### DISCUSSION TITLE
multi input from espNow

### DISCUSSION THREAD

**Discussion by aeschimannr** [pschatzmann/arduino-audio-tools#832] [Q&A] [answered]
Hello,

I managed to implement a full-duplex communication through espNow between two boards. I'm now looking to have several boards sending a stream to a single one that would mix these different inputs into a single output. 
I tried to differentiate each stream according to the mac address and redirect them to a dedicated QueueStream.
Each queueStream is then added to an InputMixer which sends it all to an i2s.

My current implementation work when i set only one queue with one emitter. But when i use 2 emitter and 2 queue i listen only strong noise.

What do you think about it? 

I2SStream i2sOut; // going to the speaker
QueueStream<uint8_t> multiQueue1(1024, 10, true); // stream from espnow 1
QueueStream<uint8_t> multiQueue2(1024, 10, true); // stream from espnow 2
InputMixer<uint8_t> mixer; 

EncodedAudioStream decoder(&multiQueue1, new SBCDecoder(256)); 
StreamCopy I2sOut(i2sOut, mixer);

TaskHandle_t TaskProcessOut;
TaskHandle_t TaskProcessIn;

// Callback when data is received
void OnDataRecv(const uint8_t * mac, const uint8_t *incomingData, int len) 
{
  char macStr[18];
  snprintf(macStr, sizeof(macStr), "%02x:%02x:%02x:%02x:%02x:%02x",
           mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
  if(strcmp(macStr, peers[0]))
  {
    // change source of the decoder
    decoder.setStream(&multiQueue1);
    // write audio data to the queue
    decoder.write(incomingData, len);
  }
  if(strcmp(macStr, peers[1]))
  {
    // change source of the decoder
    decoder.setStream(&multiQueue2);
    // write audio data to the queue
    decoder.write(incomingData, len);
  }
}

void task_processIn(void *parameter)
{
  for(;;)
  {
    mixerI2sOut.copy();
  }
}

void Setup()
{
  multiQueue1.begin();
  multiQueue2.begin();

  mixer.add(multiQueue1);
  mixer.add(multiQueue2);

  mixer.begin(config_in);

}

**Comment by pschatzmann** [comment, marked as answer]
Difficult to say what's wrong. Here a couple of guesses and proposals:
- Maybe you have some synchronization issues: after all 2 cores might be involved here. So you might need to synchronized the access or use a SynchronizedBufferRTOS.
- I would suggest to put the encoded data into the buffer. This let's you store more data and just just decode it when you need it 
- Maybe also try an other codec that is less senititive to lost data: e.g. L8

**Reply by aeschimannr** [reply]
ok, from what I understand you proposed the following scheme: 
![Screenshot 2023-05-09 at 13 56 46](https://github.com/pschatzmann/arduino-audio-tools/assets/73782115/1732fb1a-51ad-4e89-aea9-3a29066caf65)

**Reply by aeschimannr** [reply]
I have already try the L8 decoder but i wasn't able to make it work throw espnow protocol.

**Reply by aqildad-create** [reply]
> Difficult to say what's wrong. Here a couple of guesses and proposals:
> 
> * Maybe you have some synchronization issues: after all 2 cores might be involved here. So you might need to synchronized the access or use a SynchronizedBufferRTOS.
> * I would suggest to put the encoded data into the buffer. This let's you store more data and just just decode it when you need it
> * Maybe also try an other codec that is less senititive to lost data: e.g. L8

 how much audio quality will decrease with L8 FROM sbc . 

As per my knowledge SBC support 16 bit and L8 is 8bit so audio quality  will decease half ? ? ( Apology in advance if lake of knowledge  )

**Reply by pschatzmann** [reply]
If you have issues with L8, please upgrade to the library to the actual corrected version.
Here is a speed comparison of some codecs: https://www.pschatzmann.ch/home/2023/05/10/fast-audio-codecs-for-microcontrollers/
